lib/spack/spack/environment/environment.py

```diff
@@ -480,10 +480,19 @@ def specs_for_view(self, all_specs, roots):
             # recognize environment specs (which do store build deps),
             # then they need to be stripped.
             if spec.concrete:  # Do not link unconcretized roots
-                # We preserve _hash _normal to avoid recomputing DAG
-                # hashes (DAG hashes don't consider build deps)
+                # We preserve whatever the runtime hash is and _normal
+                # to avoid recomputing runtime hashes (which don't
+                # consider build deps)
                 spec_copy = spec.copy(deps=('link', 'run'))
+                # FIXME: Now the _hash (dag_hash) *does* include build deps, but
+                # if we don't copy it here, then the database will think it's not
+                # installed.
```

Comment:
Is this because we copied *just* the `(link, run)` dependencies of the spec? In that case I think this should be handled by the `Spec.copy` method and not left to client code, which might forget this crucial aspect.

Reply:
This is something that needs some light shed on it by an actual expert.  I added the `FIXME` comment because the other comments in here seemed to indicate that it was ok to strip the build deps while keeping the dag hash because the dag hash didn't consider build deps.  Now that the dag hash *does* consider build deps, is it still ok to do both of those simultaneously (strip build deps, but keep dag hash)?

Reply:
I'm going to merge this, the changes in `develop`, and https://github.com/spack/spack/pull/29634.

Reply:
The old code was really messy -- it:
1. copied the spec with only the deptypes asked for in the view
2. preserved the dag hash despite stripping build deps

@haampie's new code handles this by just traversing from roots, and it doesn't have the problems that the old `specs_for_view` did b/c it doesn't bother stripping.  So it works well with the switch to full hash and we can just keep the new linking logic from #29336.